python/cuml/experimental/common/base.pyx

```diff
@@ -0,0 +1,533 @@
+            cuml.global_settings.output_type
+            if output_type is None else output_type)
+        self._input_type = None
+        self.target_dtype = None
+        self.n_features_in_ = None
+
+        nvtx_benchmark = os.getenv('NVTX_BENCHMARK')
+        if nvtx_benchmark and nvtx_benchmark.lower() == 'true':
+            self.set_nvtx_annotations()
+
+    def __repr__(self):
+        """
+        Pretty prints the arguments of a class using Scikit-learn standard :)
+        """
+        cdef list signature = inspect.getfullargspec(self.__init__).args
+        if len(signature) > 0 and signature[0] == 'self':
+            del signature[0]
+        cdef dict state = self.__dict__
+        cdef str string = self.__class__.__name__ + '('
+        cdef str key
+        for key in signature:
+            if key not in state:
+                continue
+            if type(state[key]) is str:
+                string += "{}='{}', ".format(key, state[key])
+            else:
+                if hasattr(state[key], "__str__"):
+                    string += "{}={}, ".format(key, state[key])
+        string = string.rstrip(', ')
+        return string + ')'
+
+    def dispatch_func(self, func_name, original_func, *args, **kwargs):
+        # look for current device_type
+        device_type = cuml.global_settings.device_type
+        if device_type == 'gpu':
+            # call the original cuml method
+            return original_func(*args, **kwargs)
+        elif device_type == 'cpu':
+            # check if the sklean model already set as attribute of the cuml
+            # estimator its presence should signify that CPU execution was
+            # used previously
+            if not hasattr(self, 'sk_model_'):
+                # import model in sklearn
+                if hasattr(self, 'sk_import_path_'):
+                    # if importation path differs from the one of sklearn
+                    # look for sk_import_path_
+                    model_path = self.sk_import_path_
+                else:
+                    # importation from similar path to the current estimator
+                    # class
+                    model_path = 'sklearn' + self.__class__.__module__[4:]
+                model_name = self.__class__.__name__
+                sk_model = getattr(import_module(model_path), model_name)
+                # initialize model
+                self.sk_model_ = sk_model()
+                # transfer params set during cuml estimator initialization
+                for param in self.get_param_names():
+                    self.sk_model_.__dict__[param] = self.__dict__[param]
+
+                # transfer attributes trained with cuml
+                for attribute in self.get_attributes_names():
+                    # check presence of attribute
+                    if hasattr(self, attribute):
+                        # get the cuml attribute
+                        cu_attr = self.__dict__[attribute]
```

Comment:
What happens if there are attributes that aren't present during estimator initialization in cuML (or accessible in the state __dict__ post fit) but are needed for scikit-learn? `self.__dict__[param]` will throw an error since we don't have that param.

E.g., logistic regression `class_weight` vs `class_weight_` (and I suspect other attributes that are attached to the underlying solver model). Can this be generalized?

Reply:
If the params/attributes aren't present post-fit, then we can safely say that the algorithms are significantly different since they infer with different attributes, right?

As for the generalization of intersecting attributes, we attach the correct names in `get_attribute_names()` in each estimator we decide to support

Reply:
> If the params/attributes aren't present post-fit, then we can safely say that the algorithms are significantly different since they infer with different attributes, right?

True, but it is something we should raise an informative error, as opposed to what would be a criptic attribute not found